client/state/data-layer/wpcom-http/actions.js

```diff
@@ -30,15 +49,66 @@ export const http = ( {
 	onFailure,
 	onProgress,
 	...options,
-}, action = null ) => ( {
-	type: WPCOM_HTTP_REQUEST,
-	body,
-	method,
-	path,
-	query: { ...query, apiVersion },
-	formData,
-	onSuccess: onSuccess || action,
-	onFailure: onFailure || action,
-	onProgress: onProgress || action,
-	options,
-} );
+}, action = null ) => {
+	const actionProperties = {
+		body,
+		onSuccess: onSuccess || action,
+		onFailure: onFailure || action,
+		onProgress: onProgress || action,
+		query: { ...query },
+	};
+
+	// * check parameters *
+
+	// `path` is not optional and it is a string
+	// `method` is not optional and its value can take either 'GET' or 'POST`
+	if ( ! path || typeof path !== 'string' ) {
+		return badRequest( `'path: ${ path }' param is invalid.` );
+	}
+
+	if ( ! method || ( method !== 'GET' && method !== 'POST' ) ) {
+		return badRequest( `'method: ${ method }' param is invalid.` );
+	}
+
+	// - One of `apiVersion` of `apiNamespace` should be defined
+	if ( typeof apiVersion === 'undefined' && typeof apiNamespace === 'undefined' ) {
+		return badRequest( 'One of `apiVersion` or `apiNamespace` params must be defined.' );
+	}
+
+	// - `apiVersion` and `apiNamespace` cannot be defined simultaneously
+	if ( typeof apiVersion !== 'undefined' && typeof apiNamespace !== 'undefined' ) {
+		return badRequest( '`apiVersion` or `apiNamespace` params cannot be defined simultaneously.' );
+	}
+
+	// - `apiVersion` must have the `v1, v2, ...` shape
+	if ( typeof apiVersion !== 'undefined' ) {
+		if ( apiVersionRegex.test( apiVersion ) ) {
+			actionProperties.query.apiVersion = apiVersion;
+		} else {
+			return badRequest( `'apiVersion: ${ apiVersion }' param is invalid.` );
+		}
+	}
+
+	// - `apiNamesapce` must have the `wp/v2, wpcom/v2, ...` shape
+	if ( typeof apiNamespace !== 'undefined' ) {
+		if ( /^[a-z]+\/v\d+/.test( apiNamespace ) ) {
+			actionProperties.query.apiNamespace = apiNamespace;
+		} else {
+			return badRequest( `'apiNamespace: ${ apiNamespace }' param is invalid.` );
+		}
+	}
+
+	// optional parameters
+	if ( typeof formData !== 'undefined' ) {
+		actionProperties.formData = [];
+	}
+
+	return {
+		...actionProperties,
+		...options,
```

Comment:
@retrofox what is `foo` doing there? it shouldn't be in there. what were you trying to communicate with your code snippet?